packages/common/src/ContractInteractor.ts

```diff
@@ -940,10 +985,73 @@ calculateTransactionMaxPossibleGas: result: ${result}
     return await this.relayHubInstance.penalizer()
   }
 
+  private async _hubRelayRegistrarAddress (): Promise<Address> {
+    return await this.relayHubInstance.relayRegistrar()
+  }
+
   penalizerAddress (): Address {
     return this.penalizerInstance.address
   }
 
+  /**
+   * discover registered relays
+   * @param subset if set, then filter only to these relays
+   */
+  async getRegisteredRelays (subset?: string[], fromBlock?: number): Promise<RelayRegisteredEventInfo[]> {
+    const infoFromStorage = await this.getRegisteredRelaysFromRegistrar()
+    if (infoFromStorage != null) {
+      return infoFromStorage.filter(info => subset?.length === 0 || subset?.includes(info.relayManager))
+    } else {
+      return await this.getRegisteredRelaysFromEvents(subset, fromBlock)
+    }
+  }
+
+  async getRegisteredRelaysFromEvents (subsetManagers?: string[], fromBlock?: number): Promise<RelayRegisteredEventInfo[]> {
+    // each topic in getPastEvent is either a string or array-of-strings, to search for all.
+    const subsetManagersTopics = subsetManagers?.map(address2topic) as any as string
+    const registerEvents = await this.getPastEventsForRegistrar([subsetManagersTopics],
+      { fromBlock },
+      [RelayServerRegistered])
+    const unregisterEvents = await this.getPastEventsForStakeManager([HubUnauthorized, StakePenalized, StakeUnlocked], [subsetManagersTopics], { fromBlock })
+    // we don't check event order: removed relayer can't be re-registered, so we simply ignore any "register" of a relayer that was ever removed/unauthorized/penalized
+    const removed = new Set(unregisterEvents.map(event => event.returnValues.relayManager))
+    const relaySet: { [relayManager: string]: RelayRegisteredEventInfo } = {}
+    registerEvents
+      .filter(event => !removed.has(event.returnValues.relayManager))
+      .map(event => {
+        const { relayManager, pctRelayFee, baseRelayFee, url } = event.returnValues
+        const ret: RelayRegisteredEventInfo = {
+          relayManager, pctRelayFee, baseRelayFee, relayUrl: url
+        }
+        return ret
+      })
+      .forEach(info => {
+        relaySet[info.relayManager] = info
+      })
+    return Object.values(relaySet)
+  }
+
+  // get registered relayers from registrar
+  // (output format matches event info)
+  async getRegisteredRelaysFromRegistrar (): Promise<null | RelayRegisteredEventInfo[]> {
+    if (this.relayRegistrar == null) {
+      return null
+    }
+    const ret = await this.relayRegistrar?.readRelayInfos(0, 100)
+    const relayInfos = ret[0]
+    const filled = parseInt(ret[1].toString())
+
+    return relayInfos.slice(0, filled).map(info => {
```

Comment:
In solidity, I allocate an array for all registered relays, but then filter out unauthorized/unregistered. I didn't bother to copy the returned array again, and just returned the array (with nulls) - and also the actuall filled count.

Reply:
How bad would it be if you did copy an array again? Also, as you are not increasing the size of the array but shrinking it, isn't there a way to modify the array's length value directly? It really feels like a missing solidity feature leaking to the API and if possible that would be great to avoid it. I am not religious about it, but if that is an issue of `assembly { mstore(ret, filled)}` I think that is fine.